# 01.Importing libraries

In [1]:
# Import libraries
import pandas as pd
import numpy as np
import os

# 02.Importing data

In [29]:
# Set folder path
path = r'C:\Users\hachl\Downloads\4.3_orders_products'

In [30]:
# Import “orders_wrangled.csv” file using the os library
df_ords = pd.read_csv(os.path.join(path,'02 Data','Prepared Data','orders_wrangled.csv'), index_col = False)

In [4]:
# Import “products.csv” file using os library
df_prods = pd.read_csv(os.path.join(path,'products.csv'), index_col = False)

# 03. Data consistency checks

In [6]:
# Descriptive stats on 'df_prods'
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

## 03.1 Mixed type variables

In [7]:
# Create a new dataframe
df_test = pd.DataFrame()

In [8]:
# Create a mixed type column
df_test['mix'] = ['a', 'b', 1, True]

In [9]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [10]:
# Check for mixed types
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [11]:
# Change data type
df_test['mix'] = df_test['mix'].astype('str')

## 03.2 Missing values

In [12]:
# Find missing values in 'df_prods'
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [14]:
# Create subset of missing values in 'product_name' column
df_nan = df_prods [df_prods['product_name'].isnull() == True]

In [15]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [16]:
# Compare the number of rows in your current dataframe with the number in your subset
df_prods.shape

(49693, 5)

In [17]:
# Create subset excluding missing values
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [18]:
df_prods_clean.shape

(49677, 5)

### Alternative way: remove missing values by overwriting 'product_name' column in 'df_prods'

df_prods.dropna(subset = [‘product_name’], inplace = True)

## 03.3 Duplicates

In [19]:
# Search for duplicates: create subset containing only duplicate rows
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [20]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [22]:
# Compare shape of dataframes before and after removing the duplicates
df_prods_clean.shape

(49677, 5)

In [24]:
# Create a new dataframe that doesn’t include the duplicates you just identified
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [25]:
df_prods_clean_no_dups.shape

(49672, 5)

# 04.Exporting data

In [26]:
# Export the new dataframe 'df_prods_clean_no_dups' as “products_checked.csv” in your “Prepared Data” folder
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'))

# TASK: CONSISTENCY CHECKS ON ORDERS DATA SET

In [32]:
# Step 2: Frequency table of 'df_ords'
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_since_last_order  
count        3.421083e+06       3.421083e+06           3.214874e+06  
mean         2.776219e+00       1.345202e+01           1.111484e+01  
std          2.046829e+00       4.226088e+00           9.206737e+00  
min          0.000000e+00       0.000000e+00           0.000000e+00  
25%          1.000000e+00       1.000000e+01           4.000000e+00  
50%          3.000000e+00       1.300000e+01           7.000000e+00  
75%          5.000000e+00       1.600000e+01           1.500000e+01  
max          6.000000e+00       2.300000e+01           3.000000e+01

### What should be investigated further? By checking the frequencies values, i notice some strange values in:
- 'order_hour_of_day' column should have a minimum of 0 and a maximum of 23 (for a total of 24) but it shows a max of 2
- 'days_since_last_order' column where the 50th percentile is more than double the max. 

In [33]:
# Step 3: Check for mixed-type data in your 'df_ords'
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

### There are no columns with mixed data types to correct.

In [35]:
# Step 5: Run a check for missing values in your 'df_ords'
df_ords.isnull().sum()

Unnamed: 0                    0
order_id                      0
user_id                       0
eval_set                      0
order_number                  0
orders_day_of_week            0
order_hour_of_day             0
days_since_last_order    206209
dtype: int64

### There are 206,209 missing info in the 'days_since_last_order' column. This could mean that the user is brand new so we don´t have any prior orders in the system.

In [37]:
# Step 6: Create subset of missing values
df_ords_NaN = df_ords [df_ords['days_since_last_order'].isnull() == True]

In [38]:
df_ords_NaN

Unnamed: 0  order_id  user_id eval_set  order_number  \
0                 0   2539329        1    prior             1   
11               11   2168274        2    prior             1   
26               26   1374495        3    prior             1   
39               39   3343014        4    prior             1   
45               45   2717275        5    prior             1   
...             ...       ...      ...      ...           ...   
3420930     3420930    969311   206205    prior             1   
3420934     3420934   3189322   206206    prior             1   
3421002     3421002   2166133   206207    prior             1   
3421019     3421019   2227043   206208    prior             1   
3421069     3421069   3154581   206209    prior             1   

         orders_day_of_week  order_hour_of_day  days_since_last_order  
0                         2                  8                    NaN  
11                        2                 11                    NaN  
26                        1                 14                    NaN  
39                        6                 11                    NaN  
45                        3                 12                    NaN  
...                     ...                ...                    ...  
3420930                   4                 12                    NaN  
3420934                   3                 18                    NaN  
3421002                   6                 19                    NaN  
3421019                   1                 15                    NaN  
3421069                   3                 11                    NaN  

[206209 rows x 8 columns]

### It seems that my theory is correct as missing values on the table have 'order_number' 1, which means it´s the first order for the user. In this case, the missing value is useful information for several users, we would not remove it. No treatment for the blank observations is needed here. 

In [40]:
# Step 7: Run a check for duplicate values in your 'df_ords'
df_ords_dups = df_ords[df_ords.duplicated()]

In [43]:
# Subset containing only duplicate rows
df_ords_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, eval_set, order_number, orders_day_of_week, order_hour_of_day, days_since_last_order]
Index: []

### There is no output so it seems that all rows within the dataframe are unique.

In [45]:
# Step 9: Export your cleaned df_prods and df_ords data as “.csv” files in your “Prepared Data” folder with appropriate names
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_checked.csv'))

### I previously saved the cleaned df_products in '04. Exporting data' header as 'products_checked.csv'
### I didn´t overwrite the orders dataset but decided to save a copy as 'orders_checked.csv' to keep track of it in the same folder.